Task 1: Load 3 Months of Data for Two Thai Stocks, e.g. PTT PCL. (PTT.BK) and Bangchak Corporation (BCP.BK), both key players in Thailand’s energy sector.
Use yfinance to download daily stock prices for the past 3 months.

In [42]:
import yfinance as yf
import pandas as pd
import numpy as np

# Define stock tickers
stocks = ["KBANK.BK", "BAY.BK"]

# Download 3 months of daily data
df = yf.download(stocks, period="3mo")

# Display first few rows
df


[*********************100%***********************]  2 of 2 completed


Price           Close                High                 Low           \
Ticker         BAY.BK KBANK.BK     BAY.BK KBANK.BK     BAY.BK KBANK.BK   
Date                                                                     
2024-11-14  24.900000    146.0  25.000000    147.5  24.799999    145.5   
2024-11-15  24.700001    147.5  25.000000    148.0  24.600000    146.0   
2024-11-18  24.900000    149.0  24.900000    149.5  24.700001    148.0   
2024-11-19  25.250000    149.5  25.250000    151.0  24.900000    148.5   
2024-11-20  25.250000    148.5  25.250000    149.5  24.900000    147.0   
...               ...      ...        ...      ...        ...      ...   
2025-02-07  23.100000    160.0  23.400000    161.0  23.000000    158.0   
2025-02-10  23.299999    159.0  23.299999    160.0  23.100000    158.0   
2025-02-11  23.100000    158.0  23.299999    160.0  23.100000    158.0   
2025-02-13  23.000000    155.0  23.200001    160.0  22.900000    154.5   
2025-02-14  22.700001    153.5  23.000000    155.5  22.500000    153.0   

Price            Open           Volume            
Ticker         BAY.BK KBANK.BK  BAY.BK  KBANK.BK  
Date                                              
2024-11-14  24.799999    147.5  217600   9712900  
2024-11-15  24.900000    146.0  237500   7013300  
2024-11-18  24.700001    148.0  125400  11436600  
2024-11-19  24.900000    150.0  199400   9801000  
2024-11-20  24.900000    148.5   85000   7821600  
...               ...      ...     ...       ...  
2025-02-07  23.400000    160.5  215900   9085500  
2025-02-10  23.299999    159.5   74900  11391500  
2025-02-11  23.299999    159.0  189200   8678800  
2025-02-13  23.200001    159.5  412600  15517900  
2025-02-14  23.000000    155.5  267300  14832900  

[62 rows x 10 columns]

In [26]:
print(df.columns)

MultiIndex([( 'Close',   'BAY.BK'),
            ( 'Close', 'KBANK.BK'),
            (  'High',   'BAY.BK'),
            (  'High', 'KBANK.BK'),
            (   'Low',   'BAY.BK'),
            (   'Low', 'KBANK.BK'),
            (  'Open',   'BAY.BK'),
            (  'Open', 'KBANK.BK'),
            ('Volume',   'BAY.BK'),
            ('Volume', 'KBANK.BK')],
           names=['Price', 'Ticker'])


In [46]:
df1 = df['Close']
df1

Ticker,BAY.BK,KBANK.BK
Date,,
2024-11-14,24.900000,146.0
2024-11-15,24.700001,147.5
2024-11-18,24.900000,149.0
2024-11-19,25.250000,149.5
2024-11-20,25.250000,148.5
...,...,...
2025-02-07,23.100000,160.0
2025-02-10,23.299999,159.0
2025-02-11,23.100000,158.0


# ส่วนใหม่

# ส่วนใหม่

Task 2: Compute Daily Returns
Calculate daily percentage returns for both stocks.

In [28]:
# Compute daily returns
returns = df1.pct_change().dropna()

# Display first few rows
returns.head()


Ticker,BAY.BK,KBANK.BK
Date,,
2024-11-15,-0.008032,0.010274
2024-11-18,0.008097,0.010169
2024-11-19,0.014056,0.003356
2024-11-20,0.000000,-0.006689
2024-11-21,0.000000,0.003367


Task 3: Compute Key Metrics

3.1 Coefficient of Variation (CV): Measures risk per unit of return.

3.2 Sharpe Ratio: Measures return per unit of risk, Assume risk-free rate 1.5% per year (~0.00006 daily).

Note: You could use Scipy library:

from scipy.stats import kurtosis, skew

data.skew() #data is the daily return data

kurtosis(data)

In [55]:
#3.1 Coefficient of Variation (CV): Measures risk per unit of return.
sample_std_dev_KBANK = np.std(df1[('KBANK.BK')], ddof=1) # Access column using MultiIndex
sample_std_dev_BAY = np.std(df1[('BAY.BK')], ddof=1)  # Access column using MultiIndex

print(f"\nStandard Deviation_KBANK: {sample_std_dev_KBANK:.6f}")
print(f"\nmean_KBANK: {df1[('KBANK.BK')].mean():.6f}")

print(f"\nStandard Deviation_BAY: {sample_std_dev_BAY:.6f}")
print(f"\nmean_BAY: {df1[('BAY.BK')].mean():.6f}")


coeff_variation_KBANK = (sample_std_dev_KBANK / df1[('KBANK.BK')].mean()) # Access column using MultiIndex
print(f"\nCoefficient of Variation for KBANK: {coeff_variation_KBANK:.6f}%")

coeff_variation_BAY = (sample_std_dev_BAY / df1[('BAY.BK')].mean()) # Access column using MultiIndex
print(f"\nCoefficient of Variation for BEY: {coeff_variation_BAY:.6f}%")



Standard Deviation_KBANK: 4.422552

mean_KBANK: 156.080645

Standard Deviation_BAY: 0.655962

mean_BAY: 24.400806

Coefficient of Variation for KBANK: 0.028335%

Coefficient of Variation for BEY: 0.026883%


In [56]:
#3.2 Sharpe Ratio: Measures return per unit of risk, Assume risk-free rate 1.5% per year (~0.00006 daily).
# Assuming annual returns, calculate Sharpe Ratio
risk_free_rate = 0.00006  # 0.00006% risk-free rate

# Calculate the excess returns
excess_returns_KBANK = df1[('KBANK.BK')] - risk_free_rate
excess_returns_BEY = df1[('BAY.BK')] - risk_free_rate

# Calculate the Sharpe Ratio
sharpe_ratio_KBANK = excess_returns_KBANK.mean() / excess_returns_KBANK.std()
sharpe_ratio_BEY = excess_returns_BEY.mean() / excess_returns_BEY.std()

print(f"Sharpe Ratio for KBANK: {sharpe_ratio_KBANK:.2f}")
print(f"Sharpe Ratio for BEY: {sharpe_ratio_BEY:.2f}")

Sharpe Ratio for KBANK: 35.29
Sharpe Ratio for BEY: 37.20


Task 4: Compute the Shape of distibution for the two stocks:
*   Skewness: Measures return distribution symmetry.
*   Kurtosis: Measures tail risk (excess kurtosis).


In [57]:
#Task 4: Compute the Shape of distibution for the two stocks:
#Skewness: Measures return distribution symmetry.
kewness_KBANK = df1[('KBANK.BK')].skew()
print(f"\nSkewness for KBANK: {kewness_KBANK:.2f}")

skewness_BAY = df1[('BAY.BK')].skew()
print(f"\nSkewness for BAY: {skewness_BAY:.2f}")

#Kurtosis: Measures tail risk (excess kurtosis).
kurtosis_KBANK = df1[('KBANK.BK')].kurtosis()
print(f"\nKurtosis for KBANK: {kurtosis_KBANK:.2f}")

kurtosis_BAY = df1[('BAY.BK')].kurtosis() # Use the correct ticker 'BAY.BK'
print(f"\nKurtosis for BEY: {kurtosis_BAY:.2f}")
#Beware that Kurtosis in Python is calculated by a reference of zero for Normal dist, so Kurtosis > 0 is Lepto (higher peak)



Skewness for KBANK: -0.44

Skewness for BAY: -0.73

Kurtosis for KBANK: -0.79

Kurtosis for BEY: -0.26


Task 5: Business Decision Analysis
5.1 Which stock is better in terms of return-to-risk?

Higher Sharpe Ratio → better risk-adjusted return.

Lower CV → more stable performance.



5.2 Which stock has more risk in terms of distribution shape?

Skewness > 0 → More positive tail risk (good surprise).

Skewness < 0 → More negative tail risk (bad surprise).

Excess Kurtosis > 0 → High tail risk (black swan events).

Excess Kurtosis < 0 → Lower tail risk (less extreme moves).

Task 5: Business Decision Analysis 5.1 Which stock is better in terms of return-to-risk?

Higher Sharpe Ratio → better risk-adjusted return.

Lower CV → more stable performance.

**Sharpe Ratio:**

KBANK: 35.29

BAY: 37.20 (สูงกว่า → ดีกว่าในแง่ของผลตอบแทนปรับด้วยความเสี่ยง)

**Coefficient of Variation (CV):**

KBANK: 0.028335% (สูงกว่า → ค่าความเสี่ยงต่อ1หน่วยสูงกว่า)

BAY: 0.026883%

**สรุป BAY มี ผลตอบแทนต่อความเสี่ยงที่ดีกว่าและมีค่าความเสี่ยงต่อ1หน่วยต่ำกว่า KBANK**

5.2 Which stock has more risk in terms of distribution shape?

Skewness > 0 → More positive tail risk (good surprise).

Skewness < 0 → More negative tail risk (bad surprise).

Excess Kurtosis > 0 → High tail risk (black swan events).

Excess Kurtosis < 0 → Lower tail risk (less extreme moves).

**Skewness:**

KBANK: -0.44

BAY: -0.73

**Kurtosis:**

KBANK: -0.79

BAY: -0.26

**สรุป**:
BAYมีความเสี่ยงจากผลตอบแทนที่แย่ (bad surprise) มากกว่าเพราะSkewness ติดลบมากกว่า (-0.73)

KBANKมีความเสี่ยงจากเหตุการณ์สุดขั้วต่ำกว่า เพราะExcess Kurtosis ติดลบมากกว่า (-0.79)

**ดังนั้น** BAY มีความเสี่ยงจากการกระจายตัวของผลตอบแทนมากกว่าเนื่องจากมี Skewness ที่ติดลบมากกว่า (แสดงถึงโอกาสที่ผลตอบแทนจะต่ำกว่าค่าเฉลี่ยมากขึ้น)







